### PR TITLE
Better tests for alter table, and added the ability to modify a table with AUTO_INCREMENT

### PR DESCRIPTION
**Pull request by zachmu**

### CODE REVIEW

File: integration-tests/bats/auto_increment.bats [hunk 1]
```diff
@@ -570,14 +570,81 @@ SQL
...
+    [[ "${lines[0]}" =~ "pk" ]] || false
+    [[ "${lines[1]}" =~ "1" ]] || false
+    [[ "${lines[2]}" =~ "2" ]] || false
+    [[ "${lines[3]}" =~ "3" ]] || false
+    [[ "${lines[4]}" =~ "4" ]] || false
+    [[ "${lines[5]}" =~ "5" ]] || false
+    [[ "${lines[6]}" =~ "6" ]] || false
+}
+
+@test "auto_increment: alter table modify column to auto inc works" {
```
Comment (timsehn) [inline integration-tests/bats/auto_increment.bats:589]: Maybe test a remove auto increment using alter

File: go/libraries/doltcore/sqle/tables.go [hunk 2]
```diff
@@ -1507,7 +1595,20 @@ func (t *AlterableDoltTable) updateFromRoot(ctx *sql.Context, root *doltdb.RootV
...
+	if err != nil {
+		return err
+	}
+
+	updatedTable, _, err := root.GetTable(ctx, t.tableName)
+	if err != nil {
+		return err
+	}
+
+	sch, err := updatedTable.GetSchema(ctx)
```
Comment (max-hoffman) [inline go/libraries/doltcore/sqle/tables.go:1608]: i'm not sure i understand why we're going table -> root -> schema in some of these cases

File: integration-tests/bats/auto_increment.bats (hunk 1 above)
Comment (max-hoffman) [reply integration-tests/bats/auto_increment.bats:589]: also do we already test alter add primary key auto_increment?

File: go/libraries/doltcore/sqle/tables.go (hunk 2 above)
Comment (zachmu) [reply go/libraries/doltcore/sqle/tables.go:1608]: ALTER TABLE statements can have any number of clauses, which means that CreateCheck could be called after AddColumn for example. `t.sch` is out of date in this case (having been set at table creation time). So we need to get the updated version of the table and schema in the session root.

File: integration-tests/bats/auto_increment.bats (hunk 1 above)
Comment (zachmu) [reply integration-tests/bats/auto_increment.bats]: You can't specify auto_increment when adding a primary key, you have to add the key, then modify the column

File: go/libraries/doltcore/schema/col_coll.go [hunk 3]
```diff
@@ -143,6 +143,29 @@ func (cc *ColCollection) Append(cols ...Column) *ColCollection {
 	return NewColCollection(allCols...)
 }
 
+// IndexOf returns the index of the column with the name given (case-insensitive) or -1 if it's not found
+func (cc *ColCollection) IndexOf(colName string) (int, error) {
+	idx := -1
```
Comment (max-hoffman) [inline go/libraries/doltcore/schema/col_coll.go:148]: in the past you've recommended leaning on the `NameToCol` and `TagToIndex` maps

**Review by max-hoffman — APPROVED** [review]
LGTM

File: go/libraries/doltcore/sqle/tables.go [hunk 4]
```diff
@@ -1007,12 +1016,91 @@ func (t *AlterableDoltTable) ModifyColumn(ctx *sql.Context, columnName string, c
...
+		}
+
+		rowData, err := updatedTable.GetRowData(ctx)
+		if err != nil {
+			return err
+		}
+
+		// Note that we aren't calling the public PartitionRows, because it always gets the table data from the session
+		// root, which hasn't been updated yet
+		rowIter, err := partitionRows(ctx, updatedTable, t.projectedCols, sqlutil.SinglePartition{RowData: rowData})
```
[the remark this answers is not in the document]
Comment (zachmu) [reply go/libraries/doltcore/sqle/tables.go:1040]: At Andy's advice, decided to do this entirely in the SQL layer as much as possible. First draft inspected row data directly.

File: go/libraries/doltcore/schema/col_coll.go (hunk 3 above)
Comment (zachmu) [reply go/libraries/doltcore/schema/col_coll.go:148]: Those both use case-sensitive match unfortunately, makes them more trouble than they're worth from the SQL layer

Comment (max-hoffman) [reply go/libraries/doltcore/schema/col_coll.go:148]: you added `LowerNameToCol` also

**Pull request merged**